src/checksum/mod.rs

```diff
@@ -0,0 +1,161 @@
+use crc::{crc32, Hasher32};
+use std::fmt;
+use std::num::ParseIntError;
+use std::str::FromStr;
+
+// Checksum is a fast, incrementally computable, non-cryptographic
+// checksum of the contents of a map of bytes to bytes. The checksum
+// is commutative (it is not sensitive to the order of adds/removes)
+// so two maps with the same contents will have the same checksum no
+// matter in what order the keys were added (or removed/replaced).
+#[derive(PartialEq)]
+pub struct Checksum {
+    v: u32,
+}
+
+impl Checksum {
+    pub fn new() -> Checksum {
+        Checksum { v: 0 }
+    }
+
+    // Note: do not call add() if key is already in the map. Instead, call
+    // replace().
+    pub fn add(&mut self, key: &[u8], value: &[u8]) {
+        self.v ^= v(key, value)
+    }
+
+    pub fn remove(&mut self, key: &[u8], value: &[u8]) {
+        self.v ^= v(key, value)
+    }
+
+    pub fn replace(&mut self, key: &[u8], old_value: &[u8], new_value: &[u8]) {
+        self.remove(key, old_value);
+        self.add(key, new_value);
+    }
+}
+
+fn v(key: &[u8], value: &[u8]) -> u32 {
+    let mut digest = crc32::Digest::new(crc32::IEEE);
+    digest.write(key.len().to_string().as_bytes());
+    digest.write(key);
+    digest.write(value.len().to_string().as_bytes());
+    digest.write(value);
+    digest.sum32()
+}
+
+impl fmt::Display for Checksum {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:08x}", self.v)
+    }
+}
+
+impl fmt::Debug for Checksum {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:08x}", self.v)
```

Comment:
Nit: Delegate one to the other to self-document they are the same? I think you can say something like:

```
fmt::Display::fmt(self, f)
```

Reply:
done in https://github.com/rocicorp/repc/pull/193